## Target:
1. Perform some Exploratory Data Analysis (EDA) to find some patterns in the data.
2. Determine what would be the label for training supervised models.

In the data set, there are several different cases where the booster did not land successfully. Sometimes a landing was attempted but failed due to an accident
 
 <code>True Ocean</code> means the mission outcome was successfully  landed to a specific region of the ocean 
 <code>False Ocean</code> means the mission outcome was unsuccessfully landed to a specific region of the ocean.  
 
 <code>True RTLS</code> means the mission outcome was successfully  landed to a ground pad 
 <code>False RTLS</code> means the mission outcome was unsuccessfully landed to a ground pad.
 
 <code>True ASDS</code> means the mission outcome was successfully landed on  a drone ship 
 <code>False ASDS</code> means the mission outcome was unsuccessfully landed on a drone ship.

To-Do: *Convert those outcomes into Training Labels with `1` means the booster successfully landed `0` means it was unsuccessful.*

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("falcon_dataset_part_1.csv")
df.head()

FlightNumber       Date BoosterVersion  PayloadMass Orbit    LaunchSite  \
0             1   6/4/2010       Falcon 9  6123.547647   LEO  CCSFS SLC 40   
1             2  5/22/2012       Falcon 9   525.000000   LEO  CCSFS SLC 40   
2             3   3/1/2013       Falcon 9   677.000000   ISS  CCSFS SLC 40   
3             4  9/29/2013       Falcon 9   500.000000    PO   VAFB SLC 4E   
4             5  12/3/2013       Falcon 9  3170.000000   GTO  CCSFS SLC 40   

       Outcome  Flights  GridFins  Reused   Legs LandingPad  Block  \
0    None None        1     False   False  False        NaN      1   
1    None None        1     False   False  False        NaN      1   
2    None None        1     False   False  False        NaN      1   
3  False Ocean        1     False   False  False        NaN      1   
4    None None        1     False   False  False        NaN      1   

   ReusedCount Serial   Longitude   Latitude  
0            0  B0003  -80.577366  28.561857  
1            0  B0005  -80.577366  28.561857  
2            0  B0007  -80.577366  28.561857  
3            0  B1003 -120.610829  34.632093  
4            0  B1004  -80.577366  28.561857

Let's find the percentage of missing values in the columns

In [3]:
(df.isnull().sum()/df.count())*100

FlightNumber       0.000
Date               0.000
BoosterVersion     0.000
PayloadMass        0.000
Orbit              0.000
LaunchSite         0.000
Outcome            0.000
Flights            0.000
GridFins           0.000
Reused             0.000
Legs               0.000
LandingPad        40.625
Block              0.000
ReusedCount        0.000
Serial             0.000
Longitude          0.000
Latitude           0.000
dtype: float64

Only the column <code>LandingPad</code> has 40.63% missing values.
The <code>LandingPad</code> column will retain None values to represent **when landing pads were not used**

In [4]:
df.dtypes

FlightNumber        int64
Date               object
BoosterVersion     object
PayloadMass       float64
Orbit              object
LaunchSite         object
Outcome            object
Flights             int64
GridFins             bool
Reused               bool
Legs                 bool
LandingPad         object
Block               int64
ReusedCount         int64
Serial             object
Longitude         float64
Latitude          float64
dtype: object

# Task: 1 
Let's calculate the number of Launch sites and their respective laucnhes\

The data contains several Space X  launch facilities: <a href='https://en.wikipedia.org/wiki/List_of_Cape_Canaveral_and_Merritt_Island_launch_sites?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01'>Cape Canaveral Space</a> Launch Complex 40  <b>VAFB SLC 4E </b> , Vandenberg Air Force Base Space Launch Complex 4E <b>(SLC-4E)</b>, Kennedy Space Center Launch Complex 39A <b>KSC LC 39A </b>.The location of each Launch Is placed in the column <code>LaunchSite</code>


In [5]:
print("Number of Launch sites and launches", df.LaunchSite.value_counts())
print("Launch sites by percentage", df.LaunchSite.value_counts(normalize=True))

Number of Launch sites and launches CCSFS SLC 40    55
KSC LC 39A      22
VAFB SLC 4E     13
Name: LaunchSite, dtype: int64
Launch sites by percentage CCSFS SLC 40    0.611111
KSC LC 39A      0.244444
VAFB SLC 4E     0.144444
Name: LaunchSite, dtype: float64


### TASK 2: Calculate the number and occurrence of each orbit
Let's take a look at the number of Orbits that are present in the dataset

*   <b>LEO</b>: Low Earth orbit (LEO)is an Earth-centred orbit with an altitude of 2,000 km (1,200 mi) or less (approximately one-third of the radius of Earth),\[1] or with at least 11.25 periods per day (an orbital period of 128 minutes or less) and an eccentricity less than 0.25.\[2] Most of the manmade objects in outer space are in LEO <a href='https://en.wikipedia.org/wiki/Low_Earth_orbit?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01'>\[1]</a>.

*   <b>VLEO</b>: Very Low Earth Orbits (VLEO) can be defined as the orbits with a mean altitude below 450 km. Operating in these orbits can provide a number of benefits to Earth observation spacecraft as the spacecraft operates closer to the observation<a href='https://www.researchgate.net/publication/271499606_Very_Low_Earth_Orbit_mission_concepts_for_Earth_Observation_Benefits_and_challenges?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01'>\[2]</a>.

*   <b>GTO</b> A geosynchronous orbit is a high Earth orbit that allows satellites to match Earth's rotation. Located at 22,236 miles (35,786 kilometers) above Earth's equator, this position is a valuable spot for monitoring weather, communications and surveillance. Because the satellite orbits at the same speed that the Earth is turning, the satellite seems to stay in place over a single longitude, though it may drift north to south,” NASA wrote on its Earth Observatory website <a  href="https://www.space.com/29222-geosynchronous-orbit.html?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01" >\[3] </a>.

*   <b>SSO (or SO)</b>: It is a Sun-synchronous orbit  also called a heliosynchronous orbit is a nearly polar orbit around a planet, in which the satellite passes over any given point of the planet's surface at the same local mean solar time <a href="https://en.wikipedia.org/wiki/Sun-synchronous_orbit?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01">\[4] <a>.

*   <b>ES-L1 </b>:At the Lagrange points the gravitational forces of the two large bodies cancel out in such a way that a small object placed in orbit there is in equilibrium relative to the center of mass of the large bodies. L1 is one such point between the sun and the earth <a href="https://en.wikipedia.org/wiki/Lagrange_point?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01#L1_point">\[5]</a> .

*   <b>HEO</b> A highly elliptical orbit, is an elliptic orbit with high eccentricity, usually referring to one around Earth <a href="https://en.wikipedia.org/wiki/Highly_elliptical_orbit?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01">\[6]</a>.

*   <b> ISS </b> A modular space station (habitable artificial satellite) in low Earth orbit. It is a multinational collaborative project between five participating space agencies: NASA (United States), Roscosmos (Russia), JAXA (Japan), ESA (Europe), and CSA (Canada)<a href="https://en.wikipedia.org/wiki/International_Space_Station?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01"> \[7] </a>

*   <b> MEO </b> Geocentric orbits ranging in altitude from 2,000 km (1,200 mi) to just below geosynchronous orbit at 35,786 kilometers (22,236 mi). Also known as an intermediate circular orbit. These are "most commonly at 20,200 kilometers (12,600 mi), or 20,650 kilometers (12,830 mi), with an orbital period of 12 hours <a href="https://en.wikipedia.org/wiki/List_of_orbits?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01"> \[8] </a>

*   <b> HEO </b> Geocentric orbits above the altitude of geosynchronous orbit (35,786 km or 22,236 mi) <a href="https://en.wikipedia.org/wiki/List_of_orbits?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01"> \[9] </a>

*   <b> GEO </b> It is a circular geosynchronous orbit 35,786 kilometres (22,236 miles) above Earth's equator and following the direction of Earth's rotation <a href="https://en.wikipedia.org/wiki/Geostationary_orbit?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01"> \[10] </a>

*   <b> PO </b> It is one type of satellites in which a satellite passes above or nearly above both poles of the body being orbited (usually a planet such as the Earth <a href="https://en.wikipedia.org/wiki/Polar_orbit?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01"> \[11] </a>

some are shown in the following plot:

![](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/api/Images/Orbits.png)


In [6]:
Orbits = df.Orbit.value_counts()
Orbits

GTO      27
ISS      21
VLEO     14
PO        9
LEO       7
SSO       5
MEO       3
ES-L1     1
HEO       1
SO        1
GEO       1
Name: Orbit, dtype: int64

### TASK 3: Calculate the number and occurence of mission outcome per orbit type

In [7]:
launch_df = df.copy()
launch_df.head()

FlightNumber       Date BoosterVersion  PayloadMass Orbit    LaunchSite  \
0             1   6/4/2010       Falcon 9  6123.547647   LEO  CCSFS SLC 40   
1             2  5/22/2012       Falcon 9   525.000000   LEO  CCSFS SLC 40   
2             3   3/1/2013       Falcon 9   677.000000   ISS  CCSFS SLC 40   
3             4  9/29/2013       Falcon 9   500.000000    PO   VAFB SLC 4E   
4             5  12/3/2013       Falcon 9  3170.000000   GTO  CCSFS SLC 40   

       Outcome  Flights  GridFins  Reused   Legs LandingPad  Block  \
0    None None        1     False   False  False        NaN      1   
1    None None        1     False   False  False        NaN      1   
2    None None        1     False   False  False        NaN      1   
3  False Ocean        1     False   False  False        NaN      1   
4    None None        1     False   False  False        NaN      1   

   ReusedCount Serial   Longitude   Latitude  
0            0  B0003  -80.577366  28.561857  
1            0  B0005  -80.577366  28.561857  
2            0  B0007  -80.577366  28.561857  
3            0  B1003 -120.610829  34.632093  
4            0  B1004  -80.577366  28.561857

In [8]:
landing_outcomes = launch_df['Outcome'].value_counts()
landing_outcomes

True ASDS      41
None None      19
True RTLS      14
False ASDS      6
True Ocean      5
False Ocean     2
None ASDS       2
False RTLS      1
Name: Outcome, dtype: int64

<code>True Ocean</code> means the mission outcome was successfully  landed to a specific region of the ocean while <code>False Ocean</code> means the mission outcome was unsuccessfully landed to a specific region of the ocean. <code>True RTLS</code> means the mission outcome was successfully  landed to a ground pad <code>False RTLS</code> means the mission outcome was unsuccessfully landed to a ground pad.<code>True ASDS</code> means the mission outcome was successfully  landed to a drone ship <code>False ASDS</code> means the mission outcome was unsuccessfully landed to a drone ship. <code>None ASDS</code> and <code>None None</code> these represent a failure to land.

In [9]:
lo_keys = landing_outcomes.keys()
lo_keys

Index(['True ASDS', 'None None', 'True RTLS', 'False ASDS', 'True Ocean',
       'False Ocean', 'None ASDS', 'False RTLS'],
      dtype='object')

In [10]:
for i,outcome in enumerate(lo_keys):
    print(i,outcome)

0 True ASDS
1 None None
2 True RTLS
3 False ASDS
4 True Ocean
5 False Ocean
6 None ASDS
7 False RTLS


In [11]:
bad_outcomes = lo_keys[[1,3,5,6,7]]
bad_outcomes

Index(['None None', 'False ASDS', 'False Ocean', 'None ASDS', 'False RTLS'], dtype='object')

### TASK 4: Create a landing outcome label from Outcome column
Using the <code>Outcome</code>,  create a list where the element is zero if the corresponding  row  in  <code>Outcome</code> is in the set <code>bad_outcome</code>; otherwise, it's one. Then assign it to the variable <code>landing_class</code>:

In [12]:
landing_class = []
for key,value in launch_df['Outcome'].items():
    if value in bad_outcomes:
        landing_class.append(0)
    else:
        landing_class.append(1)

In [13]:
launch_df['Landing_Class'] = landing_class
launch_df[['Landing_Class']]

Landing_Class
0               0
1               0
2               0
3               0
4               0
..            ...
85              1
86              1
87              1
88              1
89              1

[90 rows x 1 columns]

*Let's see another way to do the same thing*

In [14]:
df2 = df.copy()

df2[['Landing_Outcome','Landing_Region']] = df2['Outcome'].str.split(" ", 1 , expand = True)

df2.head()

FlightNumber       Date BoosterVersion  PayloadMass Orbit    LaunchSite  \
0             1   6/4/2010       Falcon 9  6123.547647   LEO  CCSFS SLC 40   
1             2  5/22/2012       Falcon 9   525.000000   LEO  CCSFS SLC 40   
2             3   3/1/2013       Falcon 9   677.000000   ISS  CCSFS SLC 40   
3             4  9/29/2013       Falcon 9   500.000000    PO   VAFB SLC 4E   
4             5  12/3/2013       Falcon 9  3170.000000   GTO  CCSFS SLC 40   

       Outcome  Flights  GridFins  Reused   Legs LandingPad  Block  \
0    None None        1     False   False  False        NaN      1   
1    None None        1     False   False  False        NaN      1   
2    None None        1     False   False  False        NaN      1   
3  False Ocean        1     False   False  False        NaN      1   
4    None None        1     False   False  False        NaN      1   

   ReusedCount Serial   Longitude   Latitude Landing_Outcome Landing_Region  
0            0  B0003  -80.577366  28.561857            None           None  
1            0  B0005  -80.577366  28.561857            None           None  
2            0  B0007  -80.577366  28.561857            None           None  
3            0  B1003 -120.610829  34.632093           False          Ocean  
4            0  B1004  -80.577366  28.561857            None           None

In [15]:
df2['Landing_Outcome'].replace(['None','False'], 0, inplace= True)
df2['Landing_Outcome'].replace('True', 1, inplace= True)

df2

FlightNumber        Date BoosterVersion   PayloadMass Orbit    LaunchSite  \
0              1    6/4/2010       Falcon 9   6123.547647   LEO  CCSFS SLC 40   
1              2   5/22/2012       Falcon 9    525.000000   LEO  CCSFS SLC 40   
2              3    3/1/2013       Falcon 9    677.000000   ISS  CCSFS SLC 40   
3              4   9/29/2013       Falcon 9    500.000000    PO   VAFB SLC 4E   
4              5   12/3/2013       Falcon 9   3170.000000   GTO  CCSFS SLC 40   
..           ...         ...            ...           ...   ...           ...   
85            86    9/3/2020       Falcon 9  15600.000000  VLEO    KSC LC 39A   
86            87   10/6/2020       Falcon 9  15600.000000  VLEO    KSC LC 39A   
87            88  10/18/2020       Falcon 9  15600.000000  VLEO    KSC LC 39A   
88            89  10/24/2020       Falcon 9  15600.000000  VLEO  CCSFS SLC 40   
89            90   11/5/2020       Falcon 9   3681.000000   MEO  CCSFS SLC 40   

        Outcome  Flights  GridFins  Reused   Legs                LandingPad  \
0     None None        1     False   False  False                       NaN   
1     None None        1     False   False  False                       NaN   
2     None None        1     False   False  False                       NaN   
3   False Ocean        1     False   False  False                       NaN   
4     None None        1     False   False  False                       NaN   
..          ...      ...       ...     ...    ...                       ...   
85    True ASDS        2      True    True   True  5e9e3032383ecb6bb234e7ca   
86    True ASDS        3      True    True   True  5e9e3032383ecb6bb234e7ca   
87    True ASDS        6      True    True   True  5e9e3032383ecb6bb234e7ca   
88    True ASDS        3      True    True   True  5e9e3033383ecbb9e534e7cc   
89    True ASDS        1      True   False   True  5e9e3032383ecb6bb234e7ca   

    Block  ReusedCount Serial   Longitude   Latitude  Landing_Outcome  \
0       1            0  B0003  -80.577366  28.561857                0   
1       1            0  B0005  -80.577366  28.561857                0   
2       1            0  B0007  -80.577366  28.561857                0   
3       1            0  B1003 -120.610829  34.632093                0   
4       1            0  B1004  -80.577366  28.561857                0   
..    ...          ...    ...         ...        ...              ...   
85      5           12  B1060  -80.603956  28.608058                1   
86      5           13  B1058  -80.603956  28.608058                1   
87      5           12  B1051  -80.603956  28.608058                1   
88      5           12  B1060  -80.577366  28.561857                1   
89      5            8  B1062  -80.577366  28.561857                1   

   Landing_Region  
0            None  
1            None  
2            None  
3           Ocean  
4            None  
..            ...  
85           ASDS  
86           ASDS  
87           ASDS  
88           ASDS  
89           ASDS  

[90 rows x 19 columns]

In [16]:
df2["Landing_Outcome"].mean(), launch_df["Landing_Class"].mean()

(0.6666666666666666, 0.6666666666666666)

In [17]:
df2.to_csv("Wrangled_data.csv", index=False)


In [18]:
launch_df.to_csv("Wrangled_launch_data.csv", index=False)